### PR TITLE
feat: Implement ClientSideCredentialAccessBoundaryFactory

### PR DESCRIPTION
**Pull request by nbayati**
See: [go/client-side-cab-client-library-java](http://goto.google.com/client-side-cab-client-library-java)
This PR creates the ClientSideCredentialAccessBoundaryFactory class, and implements the function to fetch the intermediary token and session key from the STS endpoint. It completes the item `Fetching and parsing intermediate tokens` in the execution plan. [go/offline-cab-sdk-execution-plan](http://goto.google.com/offline-cab-sdk-execution-plan)

### CODE REVIEW

File: oauth2_http/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (lqiu96) [inline oauth2_http/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:1]: I believe this class is intended to go under the `cab-token-generator` module, not under the existing oauth2_http module, right? This looks like it's adding it to a different package under the same module.

File: oauth2_http/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (lqiu96) [inline oauth2_http/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:25]: If possible, can we move the validation logic into the builder's build() method?

File: oauth2_http/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (lqiu96) [inline oauth2_http/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:54]: nit:
```suggestion
    String TOKEN_EXCHANGE_URL_FORMAT = "https://sts.%s/v1/token";
    this.tokenExchangeEndpoint = String.format(TOKEN_EXCHANGE_URL_FORMAT, universeDomain);
```
Can you check if `TOKEN_EXCHANGE_URL_FORMAT` exists in Oauth2Utils or make it constant in this class?

File: oauth2_http/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (lqiu96) [inline oauth2_http/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:61]: qq, why catch and rethrow this IOException? Is the message from the earlier IOException not detailed enough?

Also, can you add Javadocs for this public method?

File: oauth2_http/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (lqiu96) [inline oauth2_http/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:65]: nit: Can we use Guava's [isNullorEmpty](https://guava.dev/releases/31.1-jre/api/docs/com/google/common/base/Strings.html#isNullOrEmpty(java.lang.String))?

File: oauth2_http/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (lqiu96) [inline oauth2_http/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:57]: If this is supposed to be public facing, can we add javadocs for this? Otherwise can this be package-private or private if it's intended for internal use?

File: oauth2_http/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (nbayati) [reply oauth2_http/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:1]: Ah yes, you're right. I'll create a new module and move the code there.

File: oauth2_http/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (nbayati) [reply oauth2_http/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:57]: Good catch. I believe this will be for internal use and we'll have another function wrapped around it to fetch and refresh credentials that will be public facing.

File: oauth2_http/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (nbayati) [reply oauth2_http/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:61]: Good question. I saw this was the way it was handled in `DownscopedCredentials.java` and assumed that was necessary. @lsirac might know more about it.
 https://github.com/googleapis/google-auth-library-java/blob/7c5ed2fb2b1f6ddc74f3ec4873d2915f58d7128f/oauth2_http/java/com/google/auth/oauth2/DownscopedCredentials.java#L137

File: oauth2_http/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (nbayati) [reply oauth2_http/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:25]: Sounds good, done!

File: cab-token-generator/pom.xml
Comment (lqiu96) [inline cab-token-generator/pom.xml:8]: Can you add the release-please tag: `<!-- {x-version-update:google-auth-library-parent:current} -->` so this will be bumped?

File: cab-token-generator/pom.xml
Comment (lqiu96) [inline cab-token-generator/pom.xml:12]: Add a <name> tag after this: `<name>Google Auth Library for Java - Cab Token Generator</name>`

File: cab-token-generator/pom.xml
Comment (lqiu96) [inline cab-token-generator/pom.xml:18]: I believe this is already mentioned in the parent pom and you don't need add it here.

File: cab-token-generator/src/main/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (lqiu96) [inline cab-token-generator/src/main/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:2]:
```suggestion
 * Copyright 2024 Google LLC
```

File: oauth2_http/java/com/google/auth/oauth2/CredentialAccessBoundary.java
Comment (lqiu96) [inline oauth2_http/java/com/google/auth/oauth2/CredentialAccessBoundary.java:70]: I don't think I see this being used anywhere. Does this need to change?

File: oauth2_http/java/com/google/auth/oauth2/StsTokenExchangeResponse.java
Comment (lqiu96) [inline oauth2_http/java/com/google/auth/oauth2/StsTokenExchangeResponse.java:159]: javadocs for the public method

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (aeitzman) [inline cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:50]: Nit: Javadoc comment for this class, I think it's fine to wait until a later PR though in case anything changes.

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (aeitzman) [inline cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:54]: typo - should be "access..." not "accees"

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (aeitzman) [inline cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:55]: nit: intermediateAccessToken

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (aeitzman) [inline cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:63]: nit: refreshCredentials?

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (nbayati) [reply cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:50]: Sounds good. I'll add the docs in a later PR.

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (aeitzman) [reply cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:63]: there are two separate flows here right? There is a function that needs to refresh intermediate credentials (this one), and one that refresh a downscoped credential - I think those two could have different names but both still be refreshXYZCredentials() - i.e. call this "refreshIntermediateCredentials()". Fetch implies to me that should also return the credential, which this function doesn't

**Review by aeitzman — APPROVED** [review]

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (lsirac) [inline cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:116]: Do we need to expose that there is an intermediary token here? If we don't need to here / anywhere else, I would omit this to avoid confusion.

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (lsirac) [inline cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:121]: nit: Missing @CanIgnoreReturnValue
```suggestion
    @CanIgnoreReturnValue
    public Builder setSourceCredential(GoogleCredentials sourceCredential) {
```

File: cab-token-generator/pom.xml
Comment (lsirac) [inline cab-token-generator/pom.xml:47]:
```suggestion
</project>

```

File: oauth2_http/java/com/google/auth/oauth2/OAuth2Utils.java
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/OAuth2Utils.java:73]: Not sure if it makes sense for this to be public @lqiu96

File: oauth2_http/java/com/google/auth/oauth2/StsTokenExchangeResponse.java
Comment (lsirac) [inline oauth2_http/java/com/google/auth/oauth2/StsTokenExchangeResponse.java:43]: This should be updated -- looks like its the Request docs

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (nbayati) [reply cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:50]: I think I'll merge this PR without unit tests to unblock Jiahua's PR and add the unit tests in a future PR.

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (nbayati) [reply cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:116]: Good point! I don't think we'll expose the intermediary token, I'll update the comment.

File: oauth2_http/java/com/google/auth/oauth2/StsTokenExchangeResponse.java
Comment (nbayati) [reply oauth2_http/java/com/google/auth/oauth2/StsTokenExchangeResponse.java:43]: Ah yes! 🤦  Thanks for catching this!

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (lqiu96) [inline cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:69]: Can you remind me why we are re-throwing this exception? If there is no reason, I think we should have `refreshIfExpired()` throw the exception instead of catching and then re-throw

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (lqiu96) [inline cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:73]: I think this probably should be an IllegalStateException rather than IOException. I don't think there is any IO here.

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (lqiu96) [inline cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:95]: qq, I believe above has a null check for `getAccessToken()`. Does there need to be a null check here as well?

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (lqiu96) [inline cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:94]: is `expires_in` mapped to `expirationTime`? If so, could we update the comment to make that clearer?

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (lqiu96) [inline cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:64]: Would you be able to write some quick comments that describes some of the behavior regarding the source access token and the response access token and how they form the intermediate access token?

From what I gather from the code, but it would be helpful for a comment to help confirm:
Intermediate access token is usually the STS response, unless the response doesn't have an expiration time. If not, then a new access token is created with the source access token's expiration time unless that value also doesn't exist.

What happens when both access tokens don't have an expiration time? Will this end up failing downstream? Will this just live forever?

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (lqiu96) [inline cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:112]: can we keep this private for now and update the visibility when needed? Just to give us flexibility to change the method name/ params if needed.

File: cab-token-generator/pom.xml
Comment (lqiu96) [inline cab-token-generator/pom.xml:10]:
```suggestion
    <version>1.30.0</version><!-- {x-version-update:google-auth-library-parent:current} -->
```

This version is going to change as we cut new releases of auth. Last thing before we merge will be to update this to the latest version.

File: cab-token-generator/pom.xml
Comment (lqiu96) [inline cab-token-generator/pom.xml:44]: I believe these versions can be managed by the parent pom. I think you can remove the versions here.

File: oauth2_http/java/com/google/auth/oauth2/OAuth2Utils.java
Comment (lqiu96) [reply oauth2_http/java/com/google/auth/oauth2/OAuth2Utils.java:73]: One option for this to see if we can keep this package-private and move some of these constants to STSRequestHandler. It looks like `TOKEN_TYPE_ACCESS_BOUNDARY_INTERMEDIARY_TOKEN` and `TOKEN_EXCHANGE_URL_FORMAT` are new additions could be live in the individual classes (i.e StsREquestHandler or DownscopedCredentials).

TOKEN_TYPE_ACCESS_TOKEN is used in a few places, but if it makes sense we can move them around. I don't know if `TOKEN_TYPE_ACCESS_TOKEN` applies to non-CAB use cases.

Otherwise, I think I'm ok with making it public given the module constraints we have.

File: oauth2_http/java/com/google/auth/oauth2/StsRequestHandler.java
Comment (lqiu96) [inline oauth2_http/java/com/google/auth/oauth2/StsRequestHandler.java:200]: nit: Perhaps `access_boundary_session_key` as a package-private constant and we can use this in the tests as well?

**Review by lsirac — APPROVED** [review]

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (lqiu96) [reply cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:69]: Ok, I'm fine with this for now, but I'd rather have this be captured via logs (that is probably a separate project).

CC: @zhumin8

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (lqiu96) [reply cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:87]: > we've already handled cases where sourceCredential.refreshIfExpired() fails or the access token is empty.

Those are two separate failures scenarios, right? Sorry, I should have been clearer on my above comment. The `handler.exchangeToken()` call also throws an IOException (https://github.com/googleapis/google-auth-library-java/blob/387b46315fa27424bd001625b3c53907d7a5ce1d/oauth2_http/java/com/google/auth/oauth2/StsRequestHandler.java#L96), which I just wanted to confirm if there needed any checks for.

File: oauth2_http/java/com/google/auth/oauth2/StsRequestHandler.java
Comment (lqiu96) [reply oauth2_http/java/com/google/auth/oauth2/StsRequestHandler.java:200]: Ah, the reason I was suggesting it is because I saw a test below had:
`private static final String ACCESS_BOUNDARY_SESSION_KEY = "accessBoundarySessionKey";` and assumed it was for the key (it looks like it's for the value instead).

nit: IMO, ACCESS_BOUNDARY_SESSION_KEY should reference `access_boundary_session_key`. If the constant probably be renamed to be something like ACCESS_BOUNDARY_SESSION_KEY_VALUE

File: oauth2_http/java/com/google/auth/oauth2/OAuth2Utils.java
Comment (nbayati) [reply oauth2_http/java/com/google/auth/oauth2/OAuth2Utils.java:73]: I'm going to leave this comment open and keep it as-is for now so I can merge this PR and unblock #1571. I'll address this comment in the next PR.

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (nbayati) [reply cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:95]: A null check isn't needed here. The StsTokenExchangeResponse constructor enforces a non-null accessToken using checkNotNull. An exception would have been thrown earlier if it were null.
https://github.com/googleapis/google-auth-library-java/blob/387b46315fa27424bd001625b3c53907d7a5ce1d/oauth2_http/java/com/google/auth/oauth2/StsTokenExchangeResponse.java#L55-L62

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (nbayati) [reply cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:112]: As we discussed, I'm marking this comment as resolved since this PR is not merging to the main branch right now.

File: cab-token-generator/pom.xml
Comment (nbayati) [reply cab-token-generator/pom.xml:10]: I'll leave this comment open as a reminder to update the version before the feature is merged to the main branch.

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (nbayati) [reply cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:87]: Yes, you're right. However, since refreshCredentials() already declares throws IOException, we don't need an explicit check around exchangeToken(). The exception will propagate up automatically and the consuming code can handle it.

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (nbayati) [reply cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:64]: Added documentation to the function. Please let me know if additional clarification is needed.

Comment (lqiu96) [reply cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:80]: Thanks, that makes sense! Is there any issue if the intermediary token doesn't have an expiration time?

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (lqiu96) [inline cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:194]: qq, does CAB plan to restrict the sourceCredentials to a subset of GoogleCredentials (i.e. only Compute or SA)?

Or can any GoogleCredential be used?

**Review by lqiu96 — APPROVED** [review]
LGTM. This PR's changes are going into the `client-side-cab` feature branch and not main. The unimplemented methods and additional changes will be addressed in future PRs.

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (nbayati) [reply cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:194]: Currently, client-side CAB only works with service account tokens. In the future we might add support for other types of tokens.

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java
Comment (nbayati) [reply cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:80]: I believe the intermediary token should still work as expected without any issues even if there's no expiration time.

Comment (lqiu96) [reply cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:80]: Sounds good, could we confirm that behavior. If it does fail, could you document that or link to docs that detail the behavior?

**Pull request merged**
